homeassistant/components/device_tracker/automatic.py

```diff
@@ -0,0 +1,170 @@
+    """Validate the configuration and return an Automatic scanner."""
+    try:
+        AutomaticDeviceScanner(config, see)
+    except IOError:
+        return False
+
+    return True
+
+
+# pylint: disable=too-many-instance-attributes
+class AutomaticDeviceScanner(object):
+    """A class representing an Automatic device."""
+
+    def __init__(self, config: dict, see) -> None:
+        """Initialize the automatic device scanner."""
+        self._client_id = config.get(CONF_CLIENT_ID)
+        self._secret = config.get(CONF_SECRET)
+        self._user_name = config.get(CONF_USERNAME)
+        self._password = config.get(CONF_PASSWORD)
+        self._devices = config.get(CONF_DEVICES, None)
+        self._headers = None
+        self._access_token = None
+        self._token_expires = dt_util.now()
+        self.last_results = {}
+        self.last_trips = {}
+        self.lock = threading.Lock()
+        self.see = see
+
+        self.scan_devices()
+
+    def scan_devices(self):
+        """Scan for new devices and return a list with found device IDs."""
+        self._update_info()
+
+        return [item['id'] for item in self.last_results]
+
+    def get_device_name(self, device):
+        """Get the device name from id."""
+        vehicle = [item['display_name'] for item in self.last_results
+                   if item['id'] == device]
+
+        return vehicle[0]
+
+    def _get_access_token(self):
+        """Get the access token from automatic."""
+        if self._access_token is None or self._token_expires <= dt_util.now():
+            data = {
+                'username': self._user_name,
+                'password': self._password,
+                'client_id': self._client_id,
+                'client_secret': self._secret,
+                'grant_type': 'password',
+                'scope': SCOPE
+            }
+
+            resp = requests.post(
+                URL_AUTHORIZE,
+                data=data)
+
+            if resp.status_code != 200:
+                raise IOError(resp.content)
+
+            json = resp.json()
+
+            self._access_token = json[ATTR_ACCESS_TOKEN]
+            self._token_expires = dt_util.now() + timedelta(
+                seconds=json[ATTR_EXPIRES_IN])
+
+        self._headers = {
+            'Authorization': 'Bearer {}'.format(self._access_token)
+        }
+
+    @Throttle(MIN_TIME_BETWEEN_SCANS)
+    def _update_info(self) -> None:
+        """Update the device info."""
+        with self.lock:
+            self._get_access_token()
+
+            _LOGGER.info('Getting device states')
+            response = requests.get(URL_VEHICLES, headers=self._headers)
+
+            if response.status_code != 200:
+                raise IOError(response.content)
+
+            self.last_results = [item for item in response.json()[ATTR_RESULTS]
+                                 if self._devices is None or item[
+                                     'display_name'] in self._devices]
```

Comment:
`self._devices` defaults to `None` which will cause an error here. I also prefer that if no whitelist is given, we just show all available devices. That will make it easier for users to get started.